wireguard/metrics.go

```diff
@@ -0,0 +1,255 @@
+		"count": len(devices),
+		"dev":   devices,
+	}).Debug("collect device metrics enumerated devices")
+
+	collector.collectDeviceMetrics(devices, m)
+	collector.collectDevicePeerMetrics(devices, m)
+}
+
+func (collector *Metrics) collectDeviceMetrics(devices []*wgtypes.Device, m chan<- prometheus.Metric) {
+	collector.logCtx.Debug("collecting wg device metrics")
+
+	for _, device := range devices {
+		l := collector.defaultLabelValues(device.PublicKey.String())
+		for k, v := range deviceMetaLabelValues(device) {
+			l[k] = v
+		}
+		collector.logCtx.WithFields(logrus.Fields{
+			"dev":    device.Name,
+			"labels": l,
+		}).Debug("iterate device")
+
+		m <- prometheus.MustNewConstMetric(
+			prometheus.NewDesc(
+				wireguardMetaFQName, wireguardMetaHelpText, nil, l,
+			),
+			prometheus.GaugeValue,
+			1,
+		)
+	}
+}
+
+func (collector *Metrics) collectDevicePeerMetrics(devices []*wgtypes.Device, m chan<- prometheus.Metric) {
+	collector.logCtx.Debug("collecting wg peer(s) metrics")
+
+	collector.logCtx.WithFields(logrus.Fields{
+		"count": len(devices),
+	}).Debug("enumerated wireguard devices")
+
+	for _, device := range devices {
+		labels := collector.defaultLabelValues(device.PublicKey.String())
+		logCtx := collector.logCtx.WithFields(logrus.Fields{
+			"key":  device.PublicKey,
+			"name": device.Name,
+		})
+		for _, peer := range device.Peers {
+			logCtx.WithFields(logrus.Fields{
+				"peer_key":      peer.PublicKey,
+				"peer_endpoint": peer.Endpoint,
+			}).Debug("collect peer metrics")
+
+			serviceLabelValues := peerServiceLabelValues(&peer)
+			for k, v := range serviceLabelValues {
+				labels[k] = v
+			}
+
+			rxd, txd := collector.calculateAndUpdateDeltas(
+				peer.PublicKey, peer.ReceiveBytes,
+				peer.TransmitBytes,
+			)
+			hs := float64(peer.LastHandshakeTime.Unix())
+			collector.logCtx.WithFields(logrus.Fields{
+				"rx_delta":     rxd,
+				"tx_delta":     txd,
+				"handshake_ts": hs,
+			}).Debug("collected peer metrics")
+
+			m <- prometheus.MustNewConstMetric(
+				prometheus.NewDesc(
+					wireguardBytesRcvdFQName, wireguardBytesRcvdHelpText, nil, labels,
+				),
+				prometheus.CounterValue,
+				rxd,
+			)
+
+			m <- prometheus.MustNewConstMetric(
+				prometheus.NewDesc(
+					wireguardBytesSentFQName, wireguardBytesSentHelpText, nil, labels,
+				),
+				prometheus.CounterValue,
+				txd,
+			)
+
+			m <- prometheus.MustNewConstMetric(
+				prometheus.NewDesc(
+					wireguardLatestHandshakeIntervalFQName, wireguardLatestHandshakeIntervalHelpText, nil, labels,
+				),
+				prometheus.GaugeValue,
+				hs,
+			)
+		}
+	}
+
+}
+
+func (collector *Metrics) calculateAndUpdateDeltas(k wgtypes.Key, rx, tx int64) (float64, float64) {
+	var rd, td float64
+
+	rc, tc := collector.peerRx[k], collector.peerTx[k]
+	collector.logCtx.WithFields(logrus.Fields{
+		"current_rx": rc,
+		"updated_rx": rx,
+		"current_tx": tc,
+		"updated_tx": tx,
+	}).Debug("read raw values")
+
+	if delta := rx - rc; rc != 0 && delta > 0 {
```

Comment:
Was thinking the rx and tx figures could be reset if wireguard is disabled/enabled, but you've covered it perfectly!

Reply:
yes, this took a lot of testing to get right. it does take at least two refreshStats to get a figure going from metrics. if I didn't do the check this way, it would've created a big delta spike on the metric readings over time